### PR TITLE
Add BatchTaskRunner to StressBench to run multiple task

### PR DESCRIPTION
**Pull request by shiqi1122**
### What changes are proposed in this pull request?

According to the discussion in PR : Add write-type parameter to StressBench, add a tool BatchTaskFRunner to run multiple test for StressBench. Currently, only the plan for MasterStressBench is completed in this request.

**link : https://github.com/Alluxio/alluxio/pull/14842

### Why are the changes needed?

The plan runner can help the operater to find out the overall benchmark of the system by executing multiple StressBench test.

### Does this PR introduce any user facing changes?

The user can now use BatchTaskRunner to run multiple test.

**the introduction for BatchTaskRunner

The possible parameters are --clients, --threads, --warmup, --base, --stop-count，--create-file-size and the BaseParameter of StressBench.

The following is an example for the input in PlanRunner:
bin/alluxio runClass alluxio.stress.cli.BatchTaskRunner MasterComprehensiveFileBatchTask --stop-count 5000 --create-file-size 0 --threads 1 --base alluxio:///stress-master-base

The comprehensive plan will run tests in the following order:
CreateFile->ListDir->ListDirLocated->GetBlockLocations->GetFileStatus->OpenFile->DeleteFile

### CODE REVIEW

File: stress/shell/src/main/java/alluxio/stress/cli/PlanRunner.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+import java.util.HashMap;
+
+/**
+ * A tool to run multiple tasks for StressBench.
+ */
+public class PlanRunner {
+  private static String[] sMasterPlan = {"MasterOpenFilePlan", "MasterGetBlockLocationsPlan",
+      "MasterGetFileStatusPlan", "MasterRenameFilePlan", "MasterListDirPlan",
+      "MasterDeleteFilePlan", "MasterComprehensiveFileTestPlan"};
```
Comment (dbw9580) [inline stress/shell/src/main/java/alluxio/stress/cli/PlanRunner.java:22]:
```suggestion
  private static final ImmutableSet<String> PLAN_NAMES = ImmutableSet.of("MasterOpenFilePlan", "MasterGetBlockLocationsPlan",
      "MasterGetFileStatusPlan", "MasterRenameFilePlan", "MasterListDirPlan",
      "MasterDeleteFilePlan", "MasterComprehensiveFileTestPlan");
```

File: stress/shell/src/main/java/alluxio/stress/cli/PlanRunner.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    plan.run(args);
+  }
+
+  private static int getPlanType(String plan) {
+    HashMap<String, Integer> map = new HashMap<>();
+    for (String s : sMasterPlan) {
+      map.put(s, 1);
+    }
+    return map.getOrDefault(plan, 0);
+  }
```
Comment (dbw9580) [inline stress/shell/src/main/java/alluxio/stress/cli/PlanRunner.java:47]: Replace with `PLAN_NAMES.contains(plan)`.

File: stress/shell/src/main/java/alluxio/stress/cli/RegisterWorkerBench.java [hunk 3]
```diff
@@ -83,6 +83,12 @@ public String getBenchDescription() {
     ));
   }
 
+  @Override
+  public String checkIfMultipleTask()
+  {
```
Comment (dbw9580) [inline stress/shell/src/main/java/alluxio/stress/cli/RegisterWorkerBench.java:88]: code style: No line break before the opening brace. https://google.github.io/styleguide/javaguide.html#s4.1.2-blocks-k-r-style
Also in other files.

**Review by dbw9580 — CHANGES_REQUESTED** [review]
Thank your for the contribution! Left some quick reviews.

File: stress/shell/src/main/java/alluxio/stress/cli/PlanRunner.java (hunk 2 above)
Comment (shiqi1122) [reply stress/shell/src/main/java/alluxio/stress/cli/PlanRunner.java]: Thank you for the review!

File: stress/shell/src/main/java/alluxio/stress/cli/MasterBenchPlan.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+            "--stop-count", String.valueOf(mParameter.mAmount),
+            "--target-throughput", String.valueOf(mTargetThroughput),
+            "--warmup", mWarmUp,
+            "--duration", mDuration,
+            "--create-file-size", mParameter.mFileSize,
+            mCluster,
+        });
+      }
+    }
+    else {
```
Comment (yyongycy) [inline stress/shell/src/main/java/alluxio/stress/cli/MasterBenchPlan.java:87]: newline necessary?

File: stress/shell/src/main/java/alluxio/stress/cli/MasterBenchPlan.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+      output(results);
+    } catch (Exception e) {
+      e.printStackTrace();
+    }
+  }
+
+  private List<String[]> getCommand() {
+    List<String[]> commands = new ArrayList<>();
+
+    if (mOperation.equals("MasterComprehensiveFileTestPlan")) {
```
Comment (yyongycy) [inline stress/shell/src/main/java/alluxio/stress/cli/MasterBenchPlan.java:70]: A little confused, there was some parameter like client type? AlluxioNative and Hdfs?

File: stress/shell/src/main/java/alluxio/stress/cli/PlanRunner.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+import com.google.common.collect.ImmutableSet;
+
+/**
+ * A tool to run multiple tasks for StressBench.
+ */
+public class PlanRunner {
+  private static final ImmutableSet<String> MASTER_PLAN_NAMES = ImmutableSet.of(
+      "MasterOpenFilePlan", "MasterGetBlockLocationsPlan",
+      "MasterGetFileStatusPlan", "MasterRenameFilePlan", "MasterListDirPlan",
+      "MasterDeleteFilePlan", "MasterComprehensiveFileTestPlan");
```
Comment (jiacheliu3) [inline stress/shell/src/main/java/alluxio/stress/cli/PlanRunner.java:23]: I guess you have not yet added these?

File: stress/shell/src/main/java/alluxio/stress/cli/MasterBenchPlan.java (hunk 4 above)
Comment (jiacheliu3) [inline stress/shell/src/main/java/alluxio/stress/cli/MasterBenchPlan.java:87]: no new line here

File: stress/shell/src/main/java/alluxio/stress/cli/BenchPlan.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.stress.cli;
+
+/**
+ * Base class for Bench plan.
+ */
+public abstract class BenchPlan {
```
Comment (jiacheliu3) [inline stress/shell/src/main/java/alluxio/stress/cli/BenchPlan.java:17]: There's a package `alluxio.job.plan` already. Maybe you don't want to reuse the word "Plan"?

File: stress/common/src/main/java/alluxio/stress/common/FileSystemParameters.java [hunk 8]
```diff
@@ -55,4 +61,19 @@ public ReadType convert(String value) {
...
   }
+
+  /**
+   * @return FileSystemWriteType of this bench
+   * Converts from String to a valid FileSystemWriteType String.
+   */
+  public static class FileSystemParametersWriteTypeConverter implements IStringConverter<String> {
+    @Override
+    public String convert(String value) {
+      if (value.equals("MUST_CACHE")  || value.equals("CACHE_THROUGH")  || value.equals("THROUGH")
```
Comment (jiacheliu3) [inline stress/common/src/main/java/alluxio/stress/common/FileSystemParameters.java:72]: Can you reuse `alluxio.client.WriteType`, checking if the string is one of the enum?

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]
Thanks for the contribution! Added some 1st round comments.

File: stress/shell/src/main/java/alluxio/stress/cli/MasterBenchPlan.java (hunk 5 above)
Comment (shiqi1122) [reply stress/shell/src/main/java/alluxio/stress/cli/MasterBenchPlan.java]: After some discussion, we wanted to make these plan to be very simple, so most of the parameters are just hardcoded in the code.

File: stress/common/src/main/java/alluxio/stress/master/MasterBenchPlanParameters.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+import alluxio.stress.Parameters;
+
+import com.beust.jcommander.Parameter;
+
+/**
+ * This holds all the parameters. All fields are public for easier json ser/de without all the
+ * getters and setters.
+ */
+public class MasterBenchPlanParameters extends Parameters {
```
[the remark this answers is not in the document]
Comment (shiqi1122) [reply stress/common/src/main/java/alluxio/stress/master/MasterBenchPlanParameters.java]: I thought maybe there will be more plan to be added in the future and there will be some different parameters with the BenchParameters. And I wanted to make it clear that many of the parameters in BenchParameters are not supported in the plan. Is it better to merge the 2 parameters?

File: stress/shell/src/main/java/alluxio/stress/cli/StressMasterBench.java [hunk 10]
```diff
@@ -116,7 +116,9 @@ public void prepare() throws Exception {
       Configuration hdfsConf = new Configuration();
       // force delete, create dirs through to UFS
       hdfsConf.set(PropertyKey.Name.USER_FILE_DELETE_UNCHECKED, "true");
-      hdfsConf.set(PropertyKey.Name.USER_FILE_WRITE_TYPE_DEFAULT, "CACHE_THROUGH");
+      if (!mParameters.mWriteType.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (shiqi1122) [reply stress/shell/src/main/java/alluxio/stress/cli/StressMasterBench.java:119]: If it is empty, the writeType will be the same with the setting in alluxio-site.properties.

File: stress/shell/src/main/java/alluxio/stress/cli/PlanRunner.java (hunk 6 above)
Comment (shiqi1122) [reply stress/shell/src/main/java/alluxio/stress/cli/PlanRunner.java]: These test are now supported. The longest and most comprehensive one is "MasterComprehensiveFileTestPlan"(now kit). Are the other kit nonsense?(Maybe these are too simple kit)

File: stress/shell/src/main/java/alluxio/stress/cli/StressMasterBench.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (jiacheliu3) [reply stress/shell/src/main/java/alluxio/stress/cli/StressMasterBench.java:119]:
```
public String mWriteType = InstancedConfiguration.defaults().get(PropertyKey.USER_FILE_WRITE_TYPE_DEFAULT);
```

**Review by dbw9580 — CHANGES_REQUESTED** [review]
Second batch of comments.

File: stress/common/src/main/java/alluxio/stress/common/FileSystemParameters.java [hunk 11]
```diff
@@ -55,4 +65,19 @@ public ReadType convert(String value) {
...
+
+  /**
+   * @return FileSystemWriteType of this bench
+   * Converts from String to a valid FileSystemWriteType String.
+   */
+  public static class FileSystemParametersWriteTypeConverter implements IStringConverter<String> {
+    @Override
+    public String convert(String value) {
+      WriteType type = WriteType.valueOf(value);
+      if (type != WriteType.NONE) {
```
Comment (jiacheliu3) [inline stress/common/src/main/java/alluxio/stress/common/FileSystemParameters.java:77]: `TRY_CACHE` is marked deprecated and should not be used

File: stress/shell/src/main/java/alluxio/stress/cli/BatchTask.java [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.stress.cli;
+
+/**
+ * Base class for batch task.
```
Comment (jiacheliu3) [inline stress/shell/src/main/java/alluxio/stress/cli/BatchTask.java:15]: can you put the definition of a batch task here?

File: stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Multiple tasks runner for Master StressBench.
+ */
+public class MasterBatchTask extends BatchTask {
+  private static final Logger LOG = LoggerFactory.getLogger(MasterBatchTask.class);
+  private static final int TARGETTHROUGHPUT = 1000000;
```
Comment (jiacheliu3) [inline stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java:34]:
```suggestion
  private static final int TARGET_THROUGHPUT = 1_000_000;
```

Comment (jiacheliu3) [reply stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java:34]: why 1million?

File: stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+  private static final String WARMUP = "0s";
+  private static final String DURATION = "21600s";
```
Comment (jiacheliu3) [inline stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java:36]: how many hours is this? better unit?

File: stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+    List<String[]> command = getCommand();
+    List<Summary> results = new ArrayList<>();
+    StressMasterBench bench = new StressMasterBench();
+    for (String[] arg : command) {
+      //error on individual task will not end the whole batch task
+      try {
+        String jsonResult = bench.run(arg);
+        Summary result = (Summary) JsonSerializable.fromJson(jsonResult);
+        results.add(result);
+      } catch (Exception e) {
```
Comment (jiacheliu3) [inline stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java:67]: There are probably two kinds of exceptions:
1. You should stop here because there's no chance to continue
2. You can log this exception and move on
Are you going to handle these separately?

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]
Added a batch of comments

File: stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java (hunk 13 above)
Comment (shiqi1122) [reply stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java:34]: I just want a very big number, and I saw in the test plan zhenhua used it was 1 million. So I used this number.

Comment (jiacheliu3) [reply stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java:34]: I see, for these heuristic numbers, if you have a reason why you put this number, pls leave a comment too.

File: stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java (hunk 15 above)
Comment (shiqi1122) [reply stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java:67]: Currently, I want to log the exception and let it move on. I am considering which kind of error should stop the whole task. If there are conditions which should result in stopping the batch task, I will add them here.

File: stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+  private static final int TARGET_THROUGHPUT  = 1_000_000;
+  private static final String WARMUP = "0s";
+  // Duration is set to a huge number to ensure that the task is finished
+  private static final String DURATION = "6h";
+  private static final String CLUSTER = "--cluster";
+
+  private String mOperation;
+
+  //store the possible operations
+  static Map<String, String> sSecondOperationMapping = ImmutableMap.<String, String>builder()
```
Comment (dbw9580) [inline stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java:44]: This is an immutable map, so can be made final. Also I think the generic type arguments do not have to be repeated.

```suggestion
  private static final Map<String, String> SECOND_OPERATION_MAPPING = ImmutableMap.builder()
```

File: stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+      .put("MasterOpenFileBatchTask", "OpenFile")
+      .put("MasterGetBlockLocationsBatchTask", "GetBlockLocations")
+      .put("MasterGetFileStatusBatchTask", "MasterGetFileStatus")
```
Comment (dbw9580) [inline stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java:47]: > `MasterGetFileStatus`

Is this a valid operation name to StressMasterBench?

File: stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java (hunk 16 above)
Comment (dbw9580) [reply stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java:44]: Since this is a task name -> operation name mapping, how about renaming it to `TASK_OPERATION_MAP`? Also leave a comment on what the keys and values are, respectively.

File: stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+      .put("MasterListDirBatchTask", "ListDir")
+      .put("MasterDeleteFileBatchTask", "DeleteFile")
+      .build();
+
+  @ParametersDelegate
+  private MasterBatchTaskParameters mParameter = new MasterBatchTaskParameters();
+
+  @Override
+  public void run(String[] args) {
+    mOperation = args[0];
```
Comment (dbw9580) [inline stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java:58]: You can have JCommander handle positional arguments: https://jcommander.org/#_main_parameter

This way JCommander can do parameter validation for you, so you wouldn't need this:

```java
throw new IllegalArgumentException(String.format("Unexpected batch task name:%s",
            mOperation));
```

**Review by dbw9580 — CHANGES_REQUESTED** [review]
Another batch of comments. We are getting there!

File: stress/common/src/main/java/alluxio/stress/master/AbstractMasterBenchParameters.java [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+      description = "The benchmark will stop after this number of paths. If -1, it is not used and "
+          + "the benchmark will stop after the duration. If this is used, duration will be "
+          + "ignored. This is typically used for creating files in preparation for another "
+          + "benchmark, since the results may not be reliable with a non-duration-based "
+          + "termination condition.")
+  public int mStopCount = STOP_COUNT_INVALID;
+
+  @Parameter(names = {"--create-file-size"},
+      description = "The size of a file for the Create op, allowed to be 0. (0, 1m, 2k, 8k, etc.)")
+  public String mCreateFileSize = "0";
```
Comment (jiacheliu3) [inline stress/common/src/main/java/alluxio/stress/master/AbstractMasterBenchParameters.java:52]: These fields are extracted from MasterBenchParameters

File: stress/common/src/main/java/alluxio/stress/master/MasterBatchTaskParameters.java [hunk 20]
```diff
@@ -11,25 +11,43 @@
...
+          + "locally.")
+  public boolean mCluster = false;
 
-  @Parameter(names = {"--threads"}, description = "the number of concurrent threads to use")
-  public int mThreads = 256;
+  @Parameter(names = {"--in-process"},
+      description = "If true, runs the task in process. Otherwise, will spawn a new process to "
+          + "execute the task",
+      hidden = true)
+  public boolean mInProcess = false;
```
Comment (jiacheliu3) [inline stress/common/src/main/java/alluxio/stress/master/MasterBatchTaskParameters.java:37]: Aren't these 2 from the `BaseParameters` class? Why do you need to define them here?

File: stress/common/src/main/java/alluxio/stress/master/MasterBatchTaskParameters.java [hunk 21]
```diff
@@ -11,25 +11,43 @@
...
 
-  @Parameter(names = {"--create-file-size"},
-      description = "The size of a file for the Create op, allowed to be 0. (0, 1m, 2k, 8k, etc.)")
-  public String mFileSize = "0";
+  /**
+   * a validator use to check the input batch task type.
+   */
+  public static class PossibleBatchTaskValidator implements IParameterValidator {
```
Comment (jiacheliu3) [inline stress/common/src/main/java/alluxio/stress/master/MasterBatchTaskParameters.java:42]: BatchTaskNameValidator

File: stress/common/src/main/java/alluxio/stress/master/MasterBatchTaskParameters.java [hunk 22]
```diff
@@ -11,25 +11,43 @@
...
+    @Override
+    public void validate(String name, String value) throws ParameterException {
+      ImmutableSet<String> masterBatchTaskNames = ImmutableSet.of(
+          "MasterComprehensiveFileBatchTask");
```
Comment (jiacheliu3) [inline stress/common/src/main/java/alluxio/stress/master/MasterBatchTaskParameters.java:46]: Move this to a static final field in `MasterBatchTaskParameters`

File: stress/common/src/main/java/alluxio/stress/master/AbstractMasterBenchParameters.java [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+
+import alluxio.stress.Parameters;
+import alluxio.stress.common.FileSystemParameters;
+
+import com.beust.jcommander.Parameter;
+
+/**
+ * This holds all the parameters shared by the MasterBench and MasterBatchTask.
+ */
+public class AbstractMasterBenchParameters extends FileSystemParameters {
```
Comment (jiacheliu3) [inline stress/common/src/main/java/alluxio/stress/master/AbstractMasterBenchParameters.java:22]: `MasterBenchBaseParameters`

File: stress/shell/src/main/java/alluxio/stress/cli/BatchTask.java [hunk 24]
```diff
@@ -20,7 +20,6 @@
    * Run the StressBench Batch task.
```
Comment (jiacheliu3) [inline stress/shell/src/main/java/alluxio/stress/cli/BatchTask.java:20]: nit: `Runs`

File: stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java [hunk 25]
```diff
@@ -47,12 +46,19 @@ public void run(String[] args, Benchmark bench) {
       throw e;
     }
 
+    if (mParameter.mWriteType.equals("ALL")) {
+      System.out.format("Parameter write-type ALL is not supported in batch task %s",
```
Comment (jiacheliu3) [inline stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java:50]: `System.err`

File: stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java [hunk 26]
```diff
@@ -47,12 +46,19 @@ public void run(String[] args, Benchmark bench) {
...
+          mParameter.mTaskName);
+      return;
+    }
+
     List<String[]> command = getCommand();
     for (String[] arg : command) {
       System.out.println("-----------------------------------------------------");
       System.out.format("Now executing command : %s on MasterStressBench...%n", arg[1]);
       // error on individual task will not end the whole batch task
```
Comment (jiacheliu3) [inline stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java:59]: // If one individual task fails, the pipeline will try to move on and finish

File: stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java [hunk 27]
```diff
@@ -98,7 +115,7 @@ private String getDescription() {
             + " alluxio:///stress-master-base",
         "$ bin/alluxio runClass alluxio.stress.cli.BatchTaskRunner"
             + " MasterComprehensiveFileBatchTask --num-files 1000 --threads 10 --create-file-size"
-            + " 1k --base alluxio:///stress-master-base",
+            + " 1k --base alluxio:///stress-master-base --warmup 0s",
```
Comment (jiacheliu3) [inline stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java:118]: Is `--warmup 0s` the preferred configuration? If so, you may want to print a warning or something if you see `mParameter.mWarmup` is not 0, and explain in the warning message.

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]
Mostly looks good to me. My comments are mostly on the namings.

File: stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java [hunk 28]
```diff
@@ -100,35 +100,12 @@ public void run(String[] args) {
...
-        if (mBaseParameter.mDistributed) {
-          command.add("--distributed");
-        }
-        if (mBaseParameter.mInProcess) {
-          command.add("--in-process");
-        }
-        if (mBaseParameter.mHelp) {
-          command.add("--h");
-        }
         commandList.add(command.toArray(new String[command.size()]));
```
Comment (jiacheliu3) [inline stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java:109]:
```suggestion
        commandList.add(command.toArray(new String[0]));
```

File: stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java (hunk 27 above)
Comment (shiqi1122) [reply stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java:126]: I added a warning. Since the file operated in warmup time will not be calculated in the final result, using warmup will make the number of operations less than expected

File: stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java [hunk 29]
```diff
@@ -0,0 +1,130 @@
...
+
+        commandList.add(command.toArray(new String[command.size()]));
+      }
+    }
+    return commandList;
+  }
+
+  private String getDescription() {
+    return String.join("\n", ImmutableList.of(
+        "BatchTaskRunner is a tool to execute pre-defined group of MasterStressBench tasks",
```
Comment (jiacheliu3) [inline stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java:117]:
```suggestion
        "BatchTaskRunner is a tool to execute a pre-defined pipeline of MasterStressBench tasks",
```

File: stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java [hunk 30]
```diff
@@ -0,0 +1,130 @@
...
+
+    List<String[]> command = getCommand();
+    for (String[] arg : command) {
+      System.out.println("-----------------------------------------------------");
+      System.out.format("Now executing command : %s on MasterStressBench...%n", arg[1]);
+      // If one individual task fails, the pipeline will try to move on and finish
+      try {
+        StressMasterBench bench = new StressMasterBench();
+        String jsonResult = bench.run(arg);
+        System.out.println("Task finished successfully. The result is following :");
```
Comment (jiacheliu3) [inline stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java:73]:
```suggestion
        System.out.format("Task %s finished successfully. The result is: %n");
```

File: stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java [hunk 31]
```diff
@@ -0,0 +1,130 @@
...
+        System.out.println(jsonResult);
+      } catch (Exception e) {
+        System.err.format("Failed to finish the %s operation%n", arg[1]);
```
Comment (jiacheliu3) [inline stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java:76]:
```suggestion
        System.err.format("Failed to finish the %s operation. Moving on to the next test in the pipeline.%n", arg[1]);
```

File: stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java [hunk 32]
```diff
@@ -0,0 +1,130 @@
...
+      }
+    }
+    System.out.println("-----------------------------------------------------");
+    System.out.println("All tasks finished. You can find the test results in the outputs above.");
+  }
+
+  private List<String[]> getCommand() {
+    List<String[]> commandList = new ArrayList<>();
+
+    if (mParameter.mTaskName.equals("MasterComprehensiveFileBatchTask")) {
```
Comment (jiacheliu3) [inline stress/shell/src/main/java/alluxio/stress/cli/MasterBatchTask.java:87]: "MasterComprehensiveFileBatchTask" can be a `private static final string COMPREHENSIVE_FILE_BATCH_TASK_NAME`

File: tests/src/test/java/alluxio/stress/BatchTaskRunnerIntegrationTest.java [hunk 33]
```diff
@@ -0,0 +1,92 @@
...
+import org.junit.Test;
+
+import java.io.ByteArrayOutputStream;
+import java.io.PrintStream;
+import java.util.List;
+import java.util.Map;
+
+public class BatchTaskRunnerIntegrationTest extends AbstractStressBenchIntegrationTest {
+  @Test
+  public void MasterIntegrationFileTest() throws Exception {
```
Comment (jiacheliu3) [inline tests/src/test/java/alluxio/stress/BatchTaskRunnerIntegrationTest.java:34]: can you add a few more tests with different parameters? For example the `--client-type` have two options. What about other read/write types? What other combinations make sense?

File: stress/common/src/test/java/alluxio/stress/BaseParametersTest.java [hunk 34]
```diff
@@ -0,0 +1,70 @@
...
+        // arguments appear in pair (8 pairs)
+        "--cluster-limit", "4",
+        "--cluster-start-delay", "5s",
+        "--profile-agent", "TestProfile",
+        "--bench-timeout", "10m",
+        "--id", "TestID",
+        "--start-ms", "1000",
+        "--java-opt", " TestOption1",
+        "--java-opt", " TestOption2",
+        // arguments appear alone
```
Comment (jiacheliu3) [inline stress/common/src/test/java/alluxio/stress/BaseParametersTest.java:40]:
```suggestion
        // keys with no values
```

File: stress/common/src/test/java/alluxio/stress/BaseParametersTest.java [hunk 35]
```diff
@@ -0,0 +1,70 @@
...
+
+public class BaseParametersTest {
+  @ParametersDelegate
+  private BaseParameters mBaseParameter = new BaseParameters();
+
+  @Test
+  public void parseParametersToArgumentString() throws Exception {
+    JCommander jc = new JCommander(this);
+    String[] inputArgs = new String[]{
+        // arguments appear in pair (8 pairs)
```
Comment (jiacheliu3) [inline stress/common/src/test/java/alluxio/stress/BaseParametersTest.java:31]:
```suggestion
        // keys with values (8 pairs)
```

File: stress/common/src/test/java/alluxio/stress/BaseParametersTest.java [hunk 36]
```diff
@@ -0,0 +1,70 @@
...
+        "--in-process",
+        "--help"
+    };
+
+    jc.parse(inputArgs);
+    List<String> outputArgs = mBaseParameter.toBatchTaskArgumentString();
+
+    assertEquals(outputArgs.size(), inputArgs.length);
+
+    // for those appear in pair (8 pair), make sure they appear in the output in certain order
```
Comment (jiacheliu3) [inline stress/common/src/test/java/alluxio/stress/BaseParametersTest.java:52]:
```suggestion
    // for those that appear in pairs, make sure they appear in the output in certain order
```

File: stress/common/src/test/java/alluxio/stress/BaseParametersTest.java [hunk 37]
```diff
@@ -0,0 +1,70 @@
...
+        if (inputArgs[2 * i].equals(outputArgs.get(j))
+            && inputArgs[2 * i + 1].equals(outputArgs.get(j + 1))) {
+          found = true;
+          break;
+        }
+      }
+      assertTrue(found);
+    }
+
+    // for those appear alone
```
Comment (jiacheliu3) [inline stress/common/src/test/java/alluxio/stress/BaseParametersTest.java:65]:
```suggestion
    // for those that have no values
```

File: stress/common/src/test/java/alluxio/stress/BaseParametersTest.java [hunk 38]
```diff
@@ -0,0 +1,70 @@
...
+import org.junit.Test;
+
+import java.util.List;
+
+public class BaseParametersTest {
+  @ParametersDelegate
+  private BaseParameters mBaseParameter = new BaseParameters();
+
+  @Test
+  public void parseParametersToArgumentString() throws Exception {
```
Comment (jiacheliu3) [inline stress/common/src/test/java/alluxio/stress/BaseParametersTest.java:28]: This test is good. Can you add more tests with different combinations, especially corner cases?

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]
Thanks for the improvements! My comment are mostly about tests.

**Review by jiacheliu3 — APPROVED** [review]
LGTM. Only some nit comments on better formatting and comments. Thanks for the contribution!

**Comment by jiacheliu3** [comment]
alluxio-bot, merge this please
